### ISSUE TITLE
Error installing in 64-bit SmartOS zones with 32-bit node

### ISSUE THREAD

**Issue by jgautier** [chrisa/node-dtrace-provider#48]
When running npm install:

```
> dtrace-provider@0.2.8 install
> node-gyp rebuild

make: Entering directory 
  ACTION binding_gyp_libusdt_target_build_libusdt .
Building libusdt for i386
gmake[1]: Entering directory 
rm -f *.gch
rm -f *.o
rm -f libusdt.a
rm -f test_usdt
rm -f test_usdt32
rm -f test_usdt64
rm -f test_mem_usage
gcc -O2 -fPIC   -c -o usdt.o usdt.c
gcc -O2 -fPIC   -c -o usdt_dof_file.o usdt_dof_file.c
gcc -O2 -fPIC -o usdt_tracepoints.o -c usdt_tracepoints_i386.s
usdt_tracepoints_i386.s: Assembler messages:
usdt_tracepoints_i386.s:25: Error: invalid instruction suffix for `push'
usdt_tracepoints_i386.s:53: Error: invalid instruction suffix for `push'
usdt_tracepoints_i386.s:65: Error: invalid instruction suffix for `push'
usdt_tracepoints_i386.s:68: Error: operand type mismatch for `jmp'
gmake[1]: *** [usdt_tracepoints.o] Error 1
```

I cloned libusdt separately and ran

```
make clean all
rm -f *.gch
rm -f *.o
rm -f libusdt.a
rm -f test_usdt
rm -f test_usdt32
rm -f test_usdt64
rm -f test_mem_usage
gcc -O2 -Wall -fPIC -m64   -c -o usdt.o usdt.c
gcc -O2 -Wall -fPIC -m64   -c -o usdt_dof_file.o usdt_dof_file.c
gcc -O2 -Wall -fPIC -m64 -o usdt_tracepoints.o -c usdt_tracepoints_x86_64.s
gcc -O2 -Wall -fPIC -m64   -c -o usdt_probe.o usdt_probe.c
gcc -O2 -Wall -fPIC -m64   -c -o usdt_dof.o usdt_dof.c
gcc -O2 -Wall -fPIC -m64   -c -o usdt_dof_sections.o usdt_dof_sections.c
rm -f libusdt.a
ar cru libusdt.a usdt.o usdt_dof_file.o usdt_tracepoints.o usdt_probe.o usdt_dof.o usdt_dof_sections.o
/bin/true libusdt.a
```

So it seems like building libusdt works fine.

**Comment by RantyDave** [comment]
Don't know if this is good news or not, but I have exactly the same problem starting from a base64 (842e6fa6-6e9b-11e5-8402-1b490459e334). Here: http://pastebin.com/a4dPCSK5

The makefile has a dodgy looking chunk at the start but I'm too tired to know what to do next :)

**Comment by davepacheco** [comment]
It's working for me on that same image to just "npm install dtrace-provider":

```
[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]# uname -v
joyent_20150818T191104Z
[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]# mdata-get sdc:image_uuid
842e6fa6-6e9b-11e5-8402-1b490459e334
[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]# which node
/opt/local/bin/node
[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]# node -pe 'process.version + " " + process.arch'
v4.1.2 x64
[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]# npm ls
/root/test-issue-48
└── (empty)

[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]# V=1 npm install dtrace-provider

> dtrace-provider@0.6.0 install /root/test-issue-48/node_modules/dtrace-provider
> node scripts/install.js

make: Entering directory '/root/test-issue-48/node_modules/dtrace-provider/build'
  LD_LIBRARY_PATH=/root/test-issue-48/node_modules/dtrace-provider/build/Release/lib.host:/root/test-issue-48/node_modules/dtrace-provider/build/Release/lib.target:$LD_LIBRARY_PATH; export LD_LIBRARY_PATH; cd ../.; sh libusdt-build.sh
Building libusdt for x86_64
gmake[1]: Entering directory '/root/test-issue-48/node_modules/dtrace-provider/libusdt'
rm -f *.gch
rm -f *.o
rm -f libusdt.a
rm -f test_usdt
rm -f test_usdt32
rm -f test_usdt64
rm -f test_mem_usage
gcc -O2 -Wall -fPIC -m64   -c -o usdt.o usdt.c
gcc -O2 -Wall -fPIC -m64   -c -o usdt_dof_file.o usdt_dof_file.c
gcc -O2 -Wall -fPIC -m64 -o usdt_tracepoints.o -c usdt_tracepoints_x86_64.s
gcc -O2 -Wall -fPIC -m64   -c -o usdt_probe.o usdt_probe.c
gcc -O2 -Wall -fPIC -m64   -c -o usdt_dof.o usdt_dof.c
gcc -O2 -Wall -fPIC -m64   -c -o usdt_dof_sections.o usdt_dof_sections.c
rm -f libusdt.a
ar cru libusdt.a usdt.o usdt_dof_file.o usdt_tracepoints.o usdt_probe.o usdt_dof.o usdt_dof_sections.o 
/bin/true libusdt.a
gmake[1]: Leaving directory '/root/test-issue-48/node_modules/dtrace-provider/libusdt'
  touch Release/obj.target/libusdt.stamp
  g++ '-DNODE_GYP_MODULE_NAME=DTraceProviderBindings' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/root/.node-gyp/4.1.2/include/node -I/root/.node-gyp/4.1.2/src -I/root/.node-gyp/4.1.2/deps/uv/include -I/root/.node-gyp/4.1.2/deps/v8/include -I../libusdt -I../node_modules/nan  -fPIC -Wall -Wextra -Wno-unused-parameter -m64 -pthreads -O3 -ffunction-sections -fdata-sections -fno-omit-frame-pointer -fno-rtti -fno-exceptions -std=gnu++0x -MMD -MF ./Release/.deps/Release/obj.target/DTraceProviderBindings/dtrace_provider.o.d.raw  -c -o Release/obj.target/DTraceProviderBindings/dtrace_provider.o ../dtrace_provider.cc
  g++ '-DNODE_GYP_MODULE_NAME=DTraceProviderBindings' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/root/.node-gyp/4.1.2/include/node -I/root/.node-gyp/4.1.2/src -I/root/.node-gyp/4.1.2/deps/uv/include -I/root/.node-gyp/4.1.2/deps/v8/include -I../libusdt -I../node_modules/nan  -fPIC -Wall -Wextra -Wno-unused-parameter -m64 -pthreads -O3 -ffunction-sections -fdata-sections -fno-omit-frame-pointer -fno-rtti -fno-exceptions -std=gnu++0x -MMD -MF ./Release/.deps/Release/obj.target/DTraceProviderBindings/dtrace_probe.o.d.raw  -c -o Release/obj.target/DTraceProviderBindings/dtrace_probe.o ../dtrace_probe.cc
  g++ '-DNODE_GYP_MODULE_NAME=DTraceProviderBindings' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/root/.node-gyp/4.1.2/include/node -I/root/.node-gyp/4.1.2/src -I/root/.node-gyp/4.1.2/deps/uv/include -I/root/.node-gyp/4.1.2/deps/v8/include -I../libusdt -I../node_modules/nan  -fPIC -Wall -Wextra -Wno-unused-parameter -m64 -pthreads -O3 -ffunction-sections -fdata-sections -fno-omit-frame-pointer -fno-rtti -fno-exceptions -std=gnu++0x -MMD -MF ./Release/.deps/Release/obj.target/DTraceProviderBindings/dtrace_argument.o.d.raw  -c -o Release/obj.target/DTraceProviderBindings/dtrace_argument.o ../dtrace_argument.cc
  g++ -shared -rdynamic -m64 -pthreads -Wl,-z,allextract  -Wl,-soname=DTraceProviderBindings.node -o Release/obj.target/DTraceProviderBindings.node -Wl,--start-group Release/obj.target/DTraceProviderBindings/dtrace_provider.o Release/obj.target/DTraceProviderBindings/dtrace_probe.o Release/obj.target/DTraceProviderBindings/dtrace_argument.o -Wl,--end-group -L/root/test-issue-48/node_modules/dtrace-provider/libusdt -l usdt
  rm -rf "Release/DTraceProviderBindings.node" && cp -af "Release/obj.target/DTraceProviderBindings.node" "Release/DTraceProviderBindings.node"
make: Leaving directory '/root/test-issue-48/node_modules/dtrace-provider/build'
dtrace-provider@0.6.0 node_modules/dtrace-provider
└── nan@2.1.0
[root@cfdfad61-f7e9-e699-86d3-81f2032c504c ~/test-issue-48]#
```

@RantyDave In your output, you're trying to build sdc-cloudapi, which is using an ancient node-bunyan, which is using an old node-dtrace-provider.  (It's installing dtrace-provider 0.2.8.  We're on 0.6.0.)  So it seems that either this is an issue with sdc-cloudapi or else there's something more subtle going on.  (In the latter case, it would greatly help if you could simplify the test case.)

**Comment by davepacheco** [comment]
I'm assuming my previous note helped resolve this, but please reopen if you're still seeing an issue.

**Comment by misterdjules** [comment]
I believe this issue is related to https://github.com/chrisa/libusdt/issues/16.

In other words, when installing node-dtrace-provider _with a x86 node binary_ on a amd64 host, libusdt is built with a x86 target architecture.

In this case, the libusdt build process tries to compile x86-specific instructions on a amd64 host, which fails.

I tested that installing a branch of node-dtrace-provider that integrates [a libusdt fix for this issue](https://github.com/chrisa/libusdt/pull/17) with a x86 node binary works on a amd64 SmartOS host.

As it seems to still be a valid issue to me, I'm reopening this issue.

**Comment by melloc** [comment]
@misterdjules Thanks for fixing this! I've merged your fix into `libusdt`, and released `node-dtrace-provider@0.8.2` with it.

**Comment by misterdjules** [comment]
Thanks!
